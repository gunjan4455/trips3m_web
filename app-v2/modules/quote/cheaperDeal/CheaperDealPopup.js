import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import config from 'appConfig';
import { PdpQuotes, CloseIcon } from 'helpers/Icon/Icon';
import { initializeChatUsingRtms, MessageType } from 'helpers/rtms';
import { currentDate, toTimeStamp } from 'helpers/DateTime';
import { generateUniqueId } from 'helpers/utils';
import { ORGANISM_VIEWED, ORGANISM_CLICKED } from 'actions/segmentEvents';

import styles from './style.cm.scss';

class CheaperDealPopUp extends PureComponent {

  static propTypes = {
    trackSegment: PropTypes.func.isRequired,
    userDetails: PropTypes.object.isRequired,
    agent: PropTypes.object.isRequired,
    tripId: PropTypes.number.isRequired,
    currentQuoteId: PropTypes.number.isRequired,
    createComment: PropTypes.func.isRequired,
    addMessage: PropTypes.func.isRequired,
    closePopup: PropTypes.func.isRequired,
    afterSend: PropTypes.func
  };

  static defaultProps = {
    afterSend: () => {}
  };

  componentDidMount() {
    this.props.trackSegment({ event: ORGANISM_VIEWED, section: 'Price Matching - Popup 2' });
  }

  componentWillUnmount() {
    if (config.useRtms && this.rtms_) {
      this.rtms_.close();
    }
  }

  async onBetterDeal() {
    this.props.trackSegment({ event: ORGANISM_CLICKED, section: 'Price Matching - Popup 2', cta: 'YES' });
    const el = document.getElementsByClassName("at_tellUs");
    el.length && el[0].click();
    await this.initializeChat();
    this.sendMessage(`Dear Agent, I am being provided with a better quote. Please give me a better deal.`);
    this.props.closePopup();
  }

  sendMessage(message) {
    const { tripId, currentQuoteId, createComment, userDetails, addMessage } = this.props;
    const { id, name, email, avatar } = userDetails;
    const messageObj = {
      title: '#chat',
      comment: message,
      commentBy: 'TRAVELLER',
      commentable_id: currentQuoteId,
      commentable_type: 'Quote',
      role: 'traveller',
      userId: id,
      user_id: id,
      userName: name,
      email,
      userPic: avatar,
      createdAt: currentDate(),
      created_at: new Date().toISOString().split('.')[0] + "+00:00",
      uId: generateUniqueId(id, toTimeStamp(new Date()), message),
      notification_sub_category: 'comment_added',
      trip_id: parseInt(tripId.toString().replace(/[^0-9\.]/g, ''), 10),
      isPending: 0,
    };
    this.currentSubscription_.sendMessage({ content: messageObj });
    createComment(tripId, currentQuoteId, messageObj);
  }

  async initializeChat() {
    if (config.useRtms && this.rtms_) {
      this.rtms_.close();
    }
    const { userDetails, addMessage, agent, currentQuoteId } = this.props;
    const { currentSubscription_, rtms_ } = await initializeChatUsingRtms({
      userId: userDetails.id,
      agentId: agent.id
    });
    this.currentSubscription_ = currentSubscription_;
    this.rtms_ = rtms_;
    this.currentSubscription_.on(MessageType.MESSAGE, function (message) {
      if (message.commentable_id === currentQuoteId) {
        addMessage(message, currentQuoteId);
      }
    }.bind(this));

    this.currentSubscription_.on(MessageType.MY_MESSAGE, function (message) {
      if (message.commentable_id === currentQuoteId) {
        addMessage(message, currentQuoteId);
      }
    }.bind(this));
  };

  closePopup() {
    this.props.trackSegment({ event: ORGANISM_CLICKED, section: 'Price Matching - Popup 2', cta: 'No' });
    this.props.closePopup();
  }

  render() {
    return (
      <div className={`pr15 pl15 pt32 pb24 sbcw at_CheaperDealPopup ${styles.popUp}`}>
        <div
          className="icon-24 absolute r15 t15"
          onClick={() => this.closePopup()}
        >
          <CloseIcon />
        </div>
        <div className="flex flexDColumn alignCenter">
          <div style={{ width: "114px", height: "81px" }} className="mb15">
            <PdpQuotes />
          </div>
          <p className="f20 fw9 mb8">Got a Cheaper Deal?</p>
          <p className="mb24 pl48 pr48 text-center">Tell us if you have a better deal and we will try to match the
            price.</p>
          <div
            className="sfcw fw7 sbc1 radius4 p10 wfull text-center mb15 at_better"
            onClick={() => this.onBetterDeal()}
          >
            <span>Yes, I Have a Better Deal</span>
          </div>
          <p
            className="fw7 pfc3 at_close"
            onClick={() => this.closePopup()}
          >
            No, I Donâ€™t
          </p>
        </div>
      </div>
    );
  }
}

export default CheaperDealPopUp;
