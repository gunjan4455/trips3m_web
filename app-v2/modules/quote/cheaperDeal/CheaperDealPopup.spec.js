import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

import CheaperDealPopup from './CheaperDealPopup.js';

configure({ adapter: new Adapter() });

const DEFAULT_PROPS = {
  "agent": {
    "id": 353152,
    "name": "Agent 353152",
    "imageUrl": "https://img.traveltriangle.com/stage/attachments/tr_a_profiles/1369/original/logo_we_care.jpg?tr=,w-300",
    "star": 4,
    "tripsSold": 393,
    "reviewsCount": 412,
    "reviewLink": "https://bookings-qa3.ttdev.in/agents/3335333135322d547261-Travel-Agency-1369",
    "email": "UserEmail_353152@gmail.com"
  },
  "currentQuoteId": 12742252,
  "tripId": 7304969,
  "userDetails": {
    "id": 1296977,
    "email": "UserEmail_1296977@gmail.com",
    "first_name": "",
    "last_name": "",
    "image": "https://img.traveltriangle.com/stage/attachments/users/1296977/original/photo.jpg",
    "latest_active_trip": {
      "destination_name": "Sikkim - Gangtok - Darjeeling",
      "trip_id": 7304969,
      "destination_image": "https://img.traveltriangle.com/stage/attachments/destinations/171/original/sgd.JPG?tr=,w-50,h-50",
      "status_txt": "Update on Quotes here",
      "trip_path": "/requested_trips/7304969-User-Email-"
    },
    "dashboard_url": "/requested_trips/7304969-User-Email-"
  },
  "trackSegment": () => {},
  "createComment": () => {},
  "addMessage": () => {},
  "closePopup": () => {},
  "afterSend": () => {}
}


const shallowWrapper = (props) => {
  return shallow(<CheaperDealPopup {...props} />);
}

describe('Should render correctly', () => {
  let props = { ...DEFAULT_PROPS };

  it('should render correctly', () => {
    const wrapper = shallowWrapper(props);
    expect(wrapper).toMatchSnapshot();
  })

  it('On Better Deal should click correctly', () => {
    const wrapper = shallowWrapper(props);
    const mockBetterDeal = jest.fn();
    wrapper.instance().onBetterDeal = mockBetterDeal;
    wrapper.find('.at_better').simulate('click');
    expect(mockBetterDeal).toHaveBeenCalled();
  });

  it('close Popup should work correctly', () => {
    const wrapper = shallowWrapper(props);
    const mockClose = jest.fn();
    wrapper.instance().closePopup = mockClose;
    wrapper.find('.at_close').simulate('click');
    expect(mockClose).toHaveBeenCalled();
  });
});
