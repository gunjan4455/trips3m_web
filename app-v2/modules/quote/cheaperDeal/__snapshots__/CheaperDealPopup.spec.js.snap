// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CheaperDealPopUp
    addMessage={[Function]}
    afterSend={[Function]}
    agent={
      Object {
        "email": "UserEmail_353152@gmail.com",
        "id": 353152,
        "imageUrl": "https://img.traveltriangle.com/stage/attachments/tr_a_profiles/1369/original/logo_we_care.jpg?tr=,w-300",
        "name": "Agent 353152",
        "reviewLink": "https://bookings-qa3.ttdev.in/agents/3335333135322d547261-Travel-Agency-1369",
        "reviewsCount": 412,
        "star": 4,
        "tripsSold": 393,
      }
    }
    closePopup={[Function]}
    createComment={[Function]}
    currentQuoteId={12742252}
    trackSegment={[Function]}
    tripId={7304969}
    userDetails={
      Object {
        "dashboard_url": "/requested_trips/7304969-User-Email-",
        "email": "UserEmail_1296977@gmail.com",
        "first_name": "",
        "id": 1296977,
        "image": "https://img.traveltriangle.com/stage/attachments/users/1296977/original/photo.jpg",
        "last_name": "",
        "latest_active_trip": Object {
          "destination_image": "https://img.traveltriangle.com/stage/attachments/destinations/171/original/sgd.JPG?tr=,w-50,h-50",
          "destination_name": "Sikkim - Gangtok - Darjeeling",
          "status_txt": "Update on Quotes here",
          "trip_id": 7304969,
          "trip_path": "/requested_trips/7304969-User-Email-",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="icon-24 absolute r15 t15"
          onClick={[Function]}
        >
          <IconSvg />
        </div>,
        <div
          className="flex flexDColumn alignCenter"
        >
          <div
            className="mb15"
            style={
              Object {
                "height": "81px",
                "width": "114px",
              }
            }
          >
            <IconSvg />
          </div>
          <p
            className="f20 fw9 mb8"
          >
            Got a Cheaper Deal?
          </p>
          <p
            className="mb24 pl48 pr48 text-center"
          >
            Tell us if you have a better deal and we will try to match the price.
          </p>
          <div
            className="sfcw fw7 sbc1 radius4 p10 wfull text-center mb15 at_better"
            onClick={[Function]}
          >
            <span>
              Yes, I Have a Better Deal
            </span>
          </div>
          <p
            className="fw7 pfc3 at_close"
            onClick={[Function]}
          >
            No, I Don’t
          </p>
        </div>,
      ],
      "className": "pr15 pl15 pt32 pb24 sbcw at_CheaperDealPopup popUp",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <IconSvg />,
          "className": "icon-24 absolute r15 t15",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="mb15"
              style={
                Object {
                  "height": "81px",
                  "width": "114px",
                }
              }
            >
              <IconSvg />
            </div>,
            <p
              className="f20 fw9 mb8"
            >
              Got a Cheaper Deal?
            </p>,
            <p
              className="mb24 pl48 pr48 text-center"
            >
              Tell us if you have a better deal and we will try to match the price.
            </p>,
            <div
              className="sfcw fw7 sbc1 radius4 p10 wfull text-center mb15 at_better"
              onClick={[Function]}
            >
              <span>
                Yes, I Have a Better Deal
              </span>
            </div>,
            <p
              className="fw7 pfc3 at_close"
              onClick={[Function]}
            >
              No, I Don’t
            </p>,
          ],
          "className": "flex flexDColumn alignCenter",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <IconSvg />,
              "className": "mb15",
              "style": Object {
                "height": "81px",
                "width": "114px",
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Got a Cheaper Deal?",
              "className": "f20 fw9 mb8",
            },
            "ref": null,
            "rendered": "Got a Cheaper Deal?",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Tell us if you have a better deal and we will try to match the price.",
              "className": "mb24 pl48 pr48 text-center",
            },
            "ref": null,
            "rendered": "Tell us if you have a better deal and we will try to match the price.",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span>
                Yes, I Have a Better Deal
              </span>,
              "className": "sfcw fw7 sbc1 radius4 p10 wfull text-center mb15 at_better",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Yes, I Have a Better Deal",
              },
              "ref": null,
              "rendered": "Yes, I Have a Better Deal",
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "No, I Don’t",
              "className": "fw7 pfc3 at_close",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "No, I Don’t",
            "type": "p",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="icon-24 absolute r15 t15"
            onClick={[Function]}
          >
            <IconSvg />
          </div>,
          <div
            className="flex flexDColumn alignCenter"
          >
            <div
              className="mb15"
              style={
                Object {
                  "height": "81px",
                  "width": "114px",
                }
              }
            >
              <IconSvg />
            </div>
            <p
              className="f20 fw9 mb8"
            >
              Got a Cheaper Deal?
            </p>
            <p
              className="mb24 pl48 pr48 text-center"
            >
              Tell us if you have a better deal and we will try to match the price.
            </p>
            <div
              className="sfcw fw7 sbc1 radius4 p10 wfull text-center mb15 at_better"
              onClick={[Function]}
            >
              <span>
                Yes, I Have a Better Deal
              </span>
            </div>
            <p
              className="fw7 pfc3 at_close"
              onClick={[Function]}
            >
              No, I Don’t
            </p>
          </div>,
        ],
        "className": "pr15 pl15 pt32 pb24 sbcw at_CheaperDealPopup popUp",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <IconSvg />,
            "className": "icon-24 absolute r15 t15",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="mb15"
                style={
                  Object {
                    "height": "81px",
                    "width": "114px",
                  }
                }
              >
                <IconSvg />
              </div>,
              <p
                className="f20 fw9 mb8"
              >
                Got a Cheaper Deal?
              </p>,
              <p
                className="mb24 pl48 pr48 text-center"
              >
                Tell us if you have a better deal and we will try to match the price.
              </p>,
              <div
                className="sfcw fw7 sbc1 radius4 p10 wfull text-center mb15 at_better"
                onClick={[Function]}
              >
                <span>
                  Yes, I Have a Better Deal
                </span>
              </div>,
              <p
                className="fw7 pfc3 at_close"
                onClick={[Function]}
              >
                No, I Don’t
              </p>,
            ],
            "className": "flex flexDColumn alignCenter",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <IconSvg />,
                "className": "mb15",
                "style": Object {
                  "height": "81px",
                  "width": "114px",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Got a Cheaper Deal?",
                "className": "f20 fw9 mb8",
              },
              "ref": null,
              "rendered": "Got a Cheaper Deal?",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Tell us if you have a better deal and we will try to match the price.",
                "className": "mb24 pl48 pr48 text-center",
              },
              "ref": null,
              "rendered": "Tell us if you have a better deal and we will try to match the price.",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span>
                  Yes, I Have a Better Deal
                </span>,
                "className": "sfcw fw7 sbc1 radius4 p10 wfull text-center mb15 at_better",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Yes, I Have a Better Deal",
                },
                "ref": null,
                "rendered": "Yes, I Have a Better Deal",
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "No, I Don’t",
                "className": "fw7 pfc3 at_close",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "No, I Don’t",
              "type": "p",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
